crates/core/tedge_mapper/src/c8y/converter.rs

```diff
@@ -144,12 +144,32 @@ where
         Ok(vec)
     }
 
-    fn try_convert_event(&mut self, input: &Message) -> Result<Vec<Message>, ConversionError> {
+    async fn try_convert_event(
```

Comment:
Having this method `async` is a sign that something is wrong.

The converter must *only* translate local messages into cloud-specific messages (or vise versa) and should not take care of sending them. The caller loop is responsible of the transport protocol, actually getting the messages and sending them.

A better design - now that we have two protocols, would be to change the return type to something along the lines:

```
enum C8yMessage {
    OverMQTT(mqtt_channel::Message)
    OverHTTP(http::Request)
}
```

That said, this is not the first place were we twist the `Converter` design. The `try_convert` method has been made `async` to process the operation. The plan is to fix that by delating the operation execution to the `agent`. The converter responsibility will be then to built the message forwarding the operation.

Reply:
Completely agree with you. This design issue has come up in some earlier conversations as well, as discussed in this [comment](https://github.com/thin-edge/thin-edge.io/pull/859#discussion_r806472940) . I had something like what you've suggested in my mind, but, I had some concerns too, which I've expressed in that comment. But, irrespective of those concerns, we all agree that, what we have today is not the right design. So, to address this, shall I create a dedicated issue to design and implement a better abstraction to decouple HTTP message conversion and transport?

Reply:
I created a discussion where you can tell you point of view on [how to decouple HTTP message conversion and transport](https://github.com/thin-edge/thin-edge.io/discussions/915).

For this PR I have concerns but more on the motivation rather than the code itself.